### ISSUE TITLE
Coordinates intersect and DataSource get_data() interface

### ISSUE THREAD

**Issue by mlshapiro** [creare-com/podpac#59]
**Description**

- Coordinates.intersect() takes optional argument `ind`. I propose we change this `return_index` to make it clear that would you are asking for is the method to return the index of the coordinates.
   - Does it make any sense to have a `CoordinateIndex()` class? This may be overengeering, and this is also a discussion for a different topic/time.
- DataSource.get_data() takes two inputs: `coordinates` and `coordinates_slice`.  `coordinates_slice` is what gets returned from `coordinates.intersect(return_index=True)`, so its really the index of coordinates. We should rename this to be an index, not a slice, so we should rename this to be one of:
  - `coordinates_index`
  - `coordinates_idx`
  - `index`

**Comment by mpu-creare** [comment]
Incidentally, we should decide what could be contained as part of the index -- and that has to be matched by the coordinates produces when `return_index=False`.

For example, `index` could be:
1. A list of slice objects
   * Strided, or just start, end
2. A boolean array
3. A flat integer array 
4. An integer array of multi-dimensional indices

Different types of datasources may only support a subset of the above. 
* I believe numpy arrays support 1-3 with no modification. 
* OpenDAP only supports 1, barely supporting strided data. 
* Asking people to support 4. essentially requires everyone writing a datasource node to implement that last half of nearest neighbor lookups. 

Perhaps we can have a flag where a user can specify which level of indexing they support in their node.

**Comment by jmilloy** [comment]
Note that sometimes coordinate.intersect does return a slice.

**Comment by mpu-creare** [comment]
Without setting the flag? Or are you saying from my list above it returns a single slice object (i.e, not as part of a list).

**Comment by jmilloy** [comment]
No, when ind=True. I thought that it returned like a list of index arrays, on for each dimension, but sometimes the index array is actually a slice, e.g. for MonotonicCoord or UniformCoord. But let me double check.

**Comment by mpu-creare** [comment]
A list of arrays, I certainly missed that one.

**Comment by jmilloy** [comment]
That Coordinate.intersect method is complicated; needs some comments and better documentation, hard to tell what it does by reading the code. But the BaseCoord.intersect method that it uses returns sometimes a slice and sometimes an index array (when ind=True). Maybe Coordinate.intersect converts those slices to index arrays, but I didn't think so. I just wanted to point that aspect out to @mlshapiro. Tthey only need to be index arrays for Coord dimensions; MonotonicCoord and UniformCoord dimensions can use slices.

**Comment by jmilloy** [comment]
As for naming, I would consider following numpy sort and argsort, where and argwhere type convention. intersect would return the intersection as a Coordinate object and argintersect would return index arrays and/or slices. In the implementation, we rename our current intersect method to `_intersect` and then it is less important what the arguments are named; users never see them.

The argument to get_data should probably be coordinate_index instead of coordinate_slice, even though it will potentially hold slices in addition to index_arrays.

**Comment by mlshapiro** [comment]
wow this is far more nuanced than i expected - thank you for explaining. @jmilloy i like the `argintersect` implementation. This makes sense, especially due to the heterogeneous nature of the returned slide/index/coordinate locator. 

I will not be touching the `intersect` method, but I will change the `get_data()` input to `coordinate_index` and I will note in the documentation that this could be index arrays or index slices, depending on the native coordinates.

**Comment by mpu-creare** [comment]
+1 for the argintersect

**Comment by mlshapiro** [comment]
@jmilloy i assume this will also be closed by #84 

data source includes new `coordinate_index` usage

**Comment by jmilloy** [comment]
@mlshapiro @mpu-creare Looking back at this, I forgot about the `argintersect` proposal.

Currently, it is implemented similar to numpy functions like [np.unique](https://docs.scipy.org/doc/numpy/reference/generated/numpy.unique.html) with a `return_indexes` argument.

It just returns the `intersection` (as a Coordinates object) by default, or you can also return `indexes`:

```
intersection = coord.intersect(other)
intersection, indexes = coord.intersect(other, return_indexes=True)
```

Please let me know if you would prefer I use `intersect` and `argintersect` instead, or if I can close this now.
